### PR TITLE
cron job to remove stale session attendance codes weekly

### PR DESCRIPTION
**Pull request by clareconstantine**
Pair programmed with @tessawiedmann 

Weekly cron job to remove attendance codes for workshop sessions where it is too late for attendance and the codes are no longer active. As of [this change](https://github.com/code-dot-org/code-dot-org/pull/21817), the codes are no longer deleted when the session ends.

### CODE REVIEW

File: cookbooks/cdo-apps/templates/default/crontab.erb [hunk 1]
```diff
@@ -96,6 +96,7 @@
       cronjob at:'0 */2 * * *', do:deploy_dir('bin', 'cron', 'teacher_applications_to_s3')
       cronjob at:'00 17 * * *', do:deploy_dir('bin', 'cron', 'stop_inactive_adhoc_instances')
       cronjob at:'0 10 * * *', do:deploy_dir('bin', 'cron', 'regional_partner_reporting')
+      cronjob at:'1 7 * * 6', do:deploy_dir('bin', 'cron', 'cleanup_workshop_attendance_codes')
```
Comment (aoby) [inline cookbooks/cdo-apps/templates/default/crontab.erb:99]: Just confirming we want this to run at 7:01am every Saturday (UTC)?

File: bin/cron/cleanup_workshop_attendance_codes [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+require 'cdo/only_one'
+
+def main
+  ten_days_ago = Time.now - 10.days
+  three_days_ago = Time.now - 3.days
+  Pd::Session.
+    where('end > ?', ten_days_ago).
+    where('end < ?', three_days_ago).
+    find_each(&:remove_code)
+end
```
Comment (aoby) [inline bin/cron/cleanup_workshop_attendance_codes]: Why between 3 and 10 days? Note we may be reinspecting sessions that have already given up codes, and we may miss older ones if we skip a week for some reason. This should be idempotent and complete. This should include a filter for `.where.not(code: nil)` and it should NOT include an upper bound (> 10 days ago). We should be able to run this any time and it should clean up everything it needs to and no more.

The logic here should be in the model, ideally with unit tests. I'd like this cronjob to simply call something like `Pd::Session.cleanup_old_codes` which will contain all the logic.

Do we care if the workshop has ended, or only the scheduled session start and end days?

Comment (aoby) [reply bin/cron/cleanup_workshop_attendance_codes]: Conceptually, the code cleanup logic doesn't need to be aware of how often it's run, so this 10 day cap is a red flag. The frequency is just a scheduling thing. Whenever it is run, it should clean up what needs to be cleaned.

File: cookbooks/cdo-apps/templates/default/crontab.erb (hunk 1 above)
Comment (tessawiedmann) [reply cookbooks/cdo-apps/templates/default/crontab.erb:99]: Yes

**Comment by clareconstantine** [comment]
[Show attendance link 48 hours in advance of the workshop until midnight the day of the workshop](https://trello.com/c/WPiSfjeT/530-show-attendance-link-48-hours-in-advance-of-the-workshop-until-midnight-the-day-of-the-workshop)

**Pull request merged**